### PR TITLE
Initialize 6 top classes.

### LINKED ISSUE #5: [Entity] Entity set up
Implement all these classes:
1. User (operatableUser, userface, superUser )
2. RoleMap (attributes: RoleAuthorityMAP , UserRoleMAP ) <Ds> 
3. Current
4. AbstractOrganization <abstract class>
5. Project <extend abstractOrg>
6. Task <extend abstractOrg>
7. UserFile <Ds>

### PR DESCRIPTION
**Pull request by KunLW**

### CODE REVIEW

**Review by hanna0601 — DISMISSED** [review]

File: src/main/java/Entity/User.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        this.userFile = userFile;
+    }
+
+
+    // abstract methods
+    public abstract UserFile getUserFile();
+
+    public abstract List<String> getAuthorityDomain();
+
+    public abstract UserFile getFile();
```
Comment (tlan4) [inline src/main/java/Entity/User.java:47]: same method as getUserFile()?

File: src/main/java/Entity/User.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    public void setUsername(String username) {
+        this.username = username;
+    }
+
+    public List<String> getRoles() {
+        return roles;
+    }
+
+    public void setRoles(List<String> roles) {
```
Comment (tlan4) [inline src/main/java/Entity/User.java:66]: add/remove roles?

File: src/main/java/Entity/User.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
+package Entity;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class User {
+    final int uid;
+    String username;
```
Comment (tlan4) [inline src/main/java/Entity/User.java:8]: Wouldn't first name/last name be needed in a company setting? maybe in userfile. Also why have username and uid? You can't have duplicate usernames or else you have to login with uid?

File: src/main/java/Entity/User.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    public void setRoles(List<String> roles) {
+        this.roles = roles;
+    }
+
+    public Department getDpt() {
+        return dpt;
+    }
+
+    public void setDpt(Department dpt) {
```
Comment (tlan4) [inline src/main/java/Entity/User.java:74]: needs to modify the department as well (remove from old, add to new)

File: src/main/java/Entity/User.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    // getters and setters
+    public int getUid() {
+        return uid;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public void setUsername(String username) {
```
Comment (tlan4) [inline src/main/java/Entity/User.java:58]: needs to change userfile as well?

File: src/main/java/Entity/UserFile.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    // CurrProjects and CurrTasks adding and removing
+    public void addCurrProject(Project project) {
+        this.currProjects.add(project);
+    }
+
+    public void addCurrTask(Task task) {
+        this.currTasks.add(task);
+    }
+
```
Comment (tlan4) [inline src/main/java/Entity/UserFile.java:46]: Why not just use remove()

**Review by tlan4 — COMMENTED** [review]
Some questions

File: src/main/java/Entity/Project.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        this.dpt = dpt;
+        this.tasks = tasks;
+        this.funds = funds;
+    }
+
+    public void addTask(Task task) {
+        this.tasks.add(task);
+    }
+
+    public boolean removeTask(String taskName) {
```
[the remark this answers is not in the document]
Comment (KunLW) [reply src/main/java/Entity/Project.java:31]: I think it is better to remove task by input the task.name not by Task itself, what do you think? .

File: src/main/java/Entity/Project.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    public void setDpt(Department dpt) {
+        this.dpt = dpt;
+    }
+
+    public List<Task> getTasks() {
+        return tasks;
+    }
+
+    public void setTasks(List<Task> tasks) {
```
[the remark this answers is not in the document]
Comment (KunLW) [reply src/main/java/Entity/Project.java:52]: add/remove tasks are already given above.

File: src/main/java/Entity/User.java (hunk 3 above)
Comment (KunLW) [reply src/main/java/Entity/User.java:8]: Sure, first name/last name make sense! We could add it. Anyway, is put it into User abstract class a good idea? Maybe is better to be put into a common User(which we have deleted, but we can add back if needed)? id is given by system and could never be changed, but username can be customised by users. Users can log in through its username or its id with correct password. So I prefer to keep it. Both username and id need to be unique. What do you think? Moreover, what kind of type for first/last name is better? Do you think this is good: two string attributes, and one for first name one for last name?

Comment (KunLW) [reply src/main/java/Entity/User.java:8]: I delete username in User and keep username in UserFile. What do you think?

Comment (tlan4) [reply src/main/java/Entity/User.java:8]: Makes sense for username, and strings for first + last name sounds good. I guess username and id makes sense like uoft.

**Review by jianwang0722 — DISMISSED** [review]

**Review by jianwang0722 — APPROVED** [review]
thanks a lot!

**Review by hanna0601 — APPROVED** [review]

**Review by tlan4 — APPROVED** [review]
👍

**Pull request merged**